src/Hazelcast.Net/Sql/ISqlQueryResult.cs

```diff
@@ -0,0 +1,56 @@
+﻿// Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+
+namespace Hazelcast.Sql
+{
+    /// <summary>
+    /// Represents SQL query (SELECT ...).
+    /// Provides methods to enumerate queried rows or cancel/dispose the query.
+    /// </summary>
+    public interface ISqlQueryResult: IAsyncEnumerator<SqlRow>, IAsyncDisposable
+    {
+        /// <summary>
+        /// <para>
+        /// Creates a one-off <see cref="IAsyncEnumerable{T}"/> around <see cref="SqlRow"/>s in row set.
+        /// </para>
+        /// <para>
+        /// Invoking this method again after enumeration has started, will throw <see cref="System.InvalidOperationException"/>.
+        /// </para>
+        /// <para>
+        /// Reusing obtained enumerable may lead to unpredictable results.
+        /// </para>
+        /// </summary>
+        /// <exception cref="ObjectDisposedException">SQL result was disposed.</exception>
+        /// <exception cref="InvalidOperationException">Enumeration has already started.</exception>
+        IAsyncEnumerable<SqlRow> EnumerateOnceAsync();
```

Comment:
Think we discussed this already so this is just for confirmation: we could make `ISqlQueryResult` an `IAsyncEnumerable<>` but the risk is that users enumerate more than once, whereas here we make it more explicit that you should enumerate only once, correct?

Reply:
In fact I am not _entirely_ convinced about that method, I find it nicer to do `await foreach (var row in result)` ... but I can understand the idea. Now one thing: we should support cancellations. That is: `await foreach (var row in result.EnumerateOnceAsync().WithCancellation(token)` - not sure this works at the moment?

Reply:
> we could make ISqlQueryResult an IAsyncEnumerable<> but the risk is that users enumerate more than once, whereas here we make it more explicit that you should enumerate only once, correct?

Yes

Reply:
Will check for cancellation support

Reply:
Added support for `WithCancellation`. Token is forwarded successfully, but there are few important notes:

- To properly support it, `SqlQueryResult` execution needs to be delayed. Before first page query started when it was constructed, now it starts with first call of `MoveNextAsync`, so that enumeration cancellation token can be used.

- It seems that server invocation itself can't be cancelled with token. So query may be cancelled in case of invocation retry or at the start/end of `MoveNextAsync`, but if page fetching itself takes too long, cancellation token won't affect it. Reason is `SendAsync` doesn't use token provided, see _MemberConnection.cs:374_. Not sure why it is implemented that way, probably worth a discussion.